libtenzir/builtins/functions/string.cpp

```diff
@@ -197,6 +197,157 @@ class nullary_method : public virtual method_plugin {
+public:
+  auto name() const -> std::string override {
+    return "replace";
+  }
+
+  auto make_function(invocation inv, session ctx) const
+    -> failure_or<function_ptr> override {
+    auto subject_expr = ast::expression{};
+    auto pattern = std::string{};
+    auto replacement = std::string{};
+    auto max_replacements = std::optional<located<int64_t>>{};
+    TRY(argument_parser2::method(name())
+          .add(subject_expr, "<string>")
+          .add(pattern, "<pattern>")
+          .add(replacement, "<pattern>")
+          .add("max", max_replacements)
+          .parse(inv, ctx));
+    if (max_replacements) {
+      if (max_replacements->inner < 0) {
+        diagnostic::error("`max` must be at least 0, but got {}",
+                          max_replacements->inner)
+          .primary(*max_replacements)
+          .emit(ctx);
+      }
+    }
+    return function_use::make(
+      [this, subject_expr = std::move(subject_expr),
+       pattern = std::move(pattern), replacement = std::move(replacement),
+       max_replacements](evaluator eval, session ctx) -> series {
+        auto result_type = string_type{};
+        auto result_arrow_type
+          = std::shared_ptr<arrow::DataType>{result_type.to_arrow_type()};
+        auto subject = eval(subject_expr);
+        auto f = detail::overload{
+          [&](const arrow::StringArray& array) {
+            auto max = max_replacements ? max_replacements->inner : -1;
+            auto options = arrow::compute::ReplaceSubstringOptions(
+              pattern, replacement, max);
+            auto result = arrow::compute::CallFunction("replace_substring",
+                                                       {array}, &options);
+            if (not result.ok()) {
+              diagnostic::warning("{}", result.status().ToString())
+                .primary(subject_expr)
+                .emit(ctx);
+              return series::null(result_type, subject.length());
+            }
+            if (not result->type()->Equals(result_arrow_type)) {
+              result = arrow::compute::Cast(result.MoveValueUnsafe(),
+                                            result_arrow_type);
+              TENZIR_ASSERT(result.ok(), result.status().ToString());
+            }
+            return series{result_type, result.MoveValueUnsafe().make_array()};
+          },
+          [&](const arrow::NullArray& array) {
+            return series::null(result_type, array.length());
+          },
+          [&](const auto&) {
+            diagnostic::warning("`{}` expected `string`, but got `{}`", name(),
+                                subject.type.kind())
+              .primary(subject_expr)
+              .emit(ctx);
+            return series::null(result_type, subject.length());
+          },
+        };
+        return caf::visit(f, *subject.array);
+      });
+  }
+};
+
+class slice : public virtual method_plugin {
+public:
+  auto name() const -> std::string override {
+    return "slice";
+  }
```

Comment:
I noted in a quick chat that we should consider making this based on graphemes instead, as code points are rarely what you'd want here. But we are fine with this for now, and might consider changing it to graphemes later (technically a breaking change).

Reply:
I agree that we want graphemes to be the proper unit. Resolving, as we don't do that right now.